source/common/io/io_uring_impl.cc

```diff
@@ -65,16 +68,63 @@ void IoUringImpl::forEveryCompletion(CompletionCb completion_cb) {
   ASSERT(SOCKET_VALID(event_fd_));
 
   eventfd_t v;
-  int ret = eventfd_read(event_fd_, &v);
-  RELEASE_ASSERT(ret == 0, "unable to drain eventfd");
+  while (true) {
+    int ret = eventfd_read(event_fd_, &v);
+    if (ret != 0) {
+      if (errno == EAGAIN) {
+        break;
+      } else {
+        ENVOY_LOG(warn, "unable to drain eventfd");
+        return;
+      }
+    }
+  }
 
   unsigned count = io_uring_peek_batch_cqe(&ring_, cqes_.data(), io_uring_size_);
 
   for (unsigned i = 0; i < count; ++i) {
     struct io_uring_cqe* cqe = cqes_[i];
     completion_cb(reinterpret_cast<void*>(cqe->user_data), cqe->res);
   }
+
   io_uring_cq_advance(&ring_, count);
+
+  ENVOY_LOG(trace, "the num of injected completion is {}", injected_completions_.size());
+
+  // TODO(soulxu): We may need to only iterate the injected completions in the current
```

Comment:
I think we should put a bound on this otherwise we could  end up watchdog killing the proxy if we have lots of cqes we're draining

Reply:
good point. that is also a reason we should iterate the injected cqes in the next event loop.

I'm also thinking of that I need to support merge events, I did see a lot of duplicated events injected.